mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala

```diff
@@ -218,7 +218,7 @@ private object IDFModel {
           newValues(k) = values(k) * idf(indices(k))
           k += 1
         }
-        Vectors.sparse(n, indices, newValues)
+        Vectors.sparse(n, indices, newValues).toSparse
```

Comment:
I can see an argument for consistent behavior with respect to explicit zeros across the three versions of `Vectors.sparse`. Sorting only happens in one of the three (as a way of detecting duplicate identifiers), and while filtering doesn't seem that expensive in comparison, it would be significant extra work in the other two versions.

The suggestion of optimizing performance with `idf.compressed` and extra cases in `transform` makes sense to me, but seems like more than is required to address this particular issue. The thought about `ArrayBuffer` suggests a reasonably small edit to the existing `SparseVector` case, though. I'll push a new commit with that change.